test_programs/compile_success_no_bug/encrypted_log_regression/src/main.nr

```diff
@@ -0,0 +1,116 @@
+global HEADER_SIZE: u32 = 48;
+global OVERHEAD_PADDING: u32 = 15;
+global OVERHEAD_SIZE: u32 = EPH_PK_SIZE + HEADER_SIZE + OVERHEAD_PADDING;
+global PLAINTEXT_LENGTH_SIZE: u32 = 2;
+global MAX_PRIVATE_LOG_PLAINTEXT_SIZE_IN_BYTES: u32 =
+    ENCRYPTED_PAYLOAD_SIZE_IN_BYTES - OVERHEAD_SIZE - PLAINTEXT_LENGTH_SIZE - 1 /* aes padding */;
+
+fn main(
+    eph_pk_bytes: [u8; EPH_PK_SIZE],
+    aes_secret: [u8; 32],
+    incoming_header_ciphertext: [u8; HEADER_SIZE],
+    extended_plaintext: [u8; MAX_PRIVATE_LOG_PLAINTEXT_SIZE_IN_BYTES + PLAINTEXT_LENGTH_SIZE],
+    shift: u32,
+) -> pub [u8; ENCRYPTED_PAYLOAD_SIZE_IN_BYTES] {
+    // unsafe {
+    //     compute_encrypted_log_unconstrained(
+    //         eph_pk_bytes,
+    //         aes_secret,
+    //         incoming_header_ciphertext,
+    //         extended_plaintext,
+    //         shift,
+    //     )
+    // }
+    compute_encrypted_log(
+        eph_pk_bytes,
+        aes_secret,
+        incoming_header_ciphertext,
+        extended_plaintext,
+        shift,
+    )
+}
+
+unconstrained fn compute_encrypted_log_unconstrained<let P: u32, let M: u32>(
+    eph_pk_bytes: [u8; EPH_PK_SIZE],
+    aes_secret: [u8; 32],
+    incoming_header_ciphertext: [u8; HEADER_SIZE],
+    plaintext: [u8; MAX_PRIVATE_LOG_PLAINTEXT_SIZE_IN_BYTES + PLAINTEXT_LENGTH_SIZE],
+    shift: u32,
+) -> [u8; M] {
+    compute_encrypted_log(
+        eph_pk_bytes,
+        aes_secret,
+        incoming_header_ciphertext,
+        plaintext,
+        shift,
+    )
+}
+
+fn compute_encrypted_log<let P: u32, let M: u32>(
+    eph_pk_bytes: [u8; EPH_PK_SIZE],
+    aes_secret: [u8; 32],
+    incoming_header_ciphertext: [u8; HEADER_SIZE],
+    plaintext: [u8; P],
+    shift: u32,
+) -> [u8; M] {
+    let mut encrypted_bytes = [0; M];
+    let mut offset = 0;
+
+    // eph_pk
+    for i in 0..EPH_PK_SIZE {
+        encrypted_bytes[offset + i] = eph_pk_bytes[i];
+    }
+    offset += EPH_PK_SIZE;
+
+    // incoming_header
+    for i in 0..HEADER_SIZE {
+        encrypted_bytes[offset + i] = incoming_header_ciphertext[i];
+    }
+    offset += HEADER_SIZE;
+
+    // Padding.
+    offset += OVERHEAD_PADDING;
+
+    // incoming_body
+    let incoming_body_ciphertext = compute_incoming_body_ciphertext(plaintext, aes_secret);
+    // Then we fill in the rest as the incoming body ciphertext
+    let size = M - offset;
+
+    // NOTE: This made the bytecode size blowup disappear in aztec packages,
+    // but in this reproduction the size seems to be statically known regardless.
+    //let size = M - 32 - HEADER_SIZE - OVERHEAD_PADDING;
+
+    assert_eq(size, incoming_body_ciphertext.len(), "ciphertext length mismatch");
+    for i in 0..size {
+        // NOTE: Adding `shift` makes the index dynamic enough to force ACIR to use memory operations.
+        // Without `shift` it assigns directly to witnesses, which is good, but not what we wanted to reproduce.
+        encrypted_bytes[offset + i + shift] = incoming_body_ciphertext[i];
```

Comment:
I have a feeling we are seeing a blow-up here because every `array_set` from this unrolled loop is performing a copy. If we see `v3 = array_set v1 index v2 value incoming_body_ciphertext[i]`, we will copy `v1` so that we do not mutate it directly.

However, in this case, we know that the dynamic value `shift` is not changed between `array_set` instructions and in fact `i` is different for each one, thus we should be able to safely mutate the array and avoid copying. Could you post the final SSA before ACIR gen? You can probably get rid of all the code preceding this loop to shorten the SSA.

Reply:
> With the shift in place the SSA ends like this

Are there `array_set` instructions in the instructions you snipped between v178 and v1944

Reply:
Hmm it does look like we are correctly marking the array sets mutable here.

Reply:
Yeah, I think we're handling this optimally here (minus the range constraints). Having a dynamic shift amount will result in this being a RAM write so this should be replaced with an unconstrained helper.

Reply:
I agree I think `shift` is a "false flag". I added it to force memory operations, which ended up resembling what the original flame graphs suggested in terms of opcode composition, but looking at the [original SSA](https://github.com/noir-lang/noir/pull/6972#issuecomment-2576190022) revealed that the range checks are coming from conditional variables, rather than indexing, so I'll try to rewrite the test to include if-then-else instead. 

I'll also check why making `size` statically known changes the original SSA to be so much smaller, because that seems unrelated to any conditionals. Maybe a known `size` gets rid of the memory operations, and then the `Remove IfElse` doesn't cause copies in a loop 🤔